server/src/service.rs

```diff
@@ -1,30 +1,77 @@
 // org.freedesktop.Secret.Service
 
-use std::collections::HashMap;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
 
-use oo7::dbus::{
-    api::{Properties, SecretInner},
-    Algorithm, ServiceError,
+use oo7::{
+    dbus::{
+        api::{Properties, SecretInner},
+        Algorithm, ServiceError,
+    },
+    Key,
 };
+use tokio::sync::RwLock;
 use zbus::{
     proxy::ProxyDefault,
     zvariant::{ObjectPath, OwnedObjectPath, OwnedValue, Value},
 };
 
+use crate::{service_manager::ServiceManager, session::Session};
+
 pub type Result<T> = std::result::Result<T, ServiceError>;
 
 #[derive(Debug, Default)]
-pub struct Service {}
+pub struct Service {
+    session_counter: RwLock<i32>,
+    manager: Arc<Mutex<ServiceManager>>,
+}
 
 #[zbus::interface(name = "org.freedesktop.Secret.Service")]
 impl Service {
     #[zbus(out_args("output", "result"))]
     pub async fn open_session(
         &self,
-        _algorithm: Algorithm,
-        _input: Value<'_>,
+        algorithm: Algorithm,
+        input: Value<'_>,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
     ) -> Result<(OwnedValue, OwnedObjectPath)> {
-        todo!()
+        let (public_key, aes_key) = match algorithm {
+            Algorithm::Plain => (None, None),
+            Algorithm::Encrypted => {
+                let client_public_key = Key::from(input);
+                let private_key = Key::generate_private_key();
+                (
+                    Some(Key::generate_public_key(&private_key)),
+                    Some(Key::generate_aes_key(&private_key, &client_public_key)),
+                )
+            }
+        };
+
+        *self.session_counter.write().await += 1;
+        let session = Session::new(
+            Some(Arc::new(aes_key.unwrap())),
+            Arc::clone(&self.manager),
+            *self.session_counter.read().await,
+        );
+        let path = session.path();
+
+        {
+            let session = Arc::new(session.clone());
+            self.manager
+                .lock()
+                .unwrap()
+                .insert_session(path.clone(), Arc::clone(&session));
+        }
+
+        object_server.at(&path, session).await?;
+
+        let service_key = public_key
+            .map(|k| OwnedValue::from(&k))
+            .unwrap_or_else(|| Value::new::<Vec<u8>>(vec![]).try_to_owned().unwrap());
```

Comment:
This looks weird.

Reply:
I had to try many times to get this right :) and I'm not sure how to make this better.

Reply:
That is okay, I will look at it after we merge this